packages/client/src/v2-events/features/events/useEvents.ts

```diff
@@ -10,76 +10,138 @@
+import { api, queryClient, utils } from '@client/v2-events/trpc'
 import { EventDocument } from '@events/schema'
-import { useQueryClient, useSuspenseQuery } from '@tanstack/react-query'
+import { useSuspenseQuery } from '@tanstack/react-query'
 
 export function preloadData() {
   utils.config.get.ensureData()
 }
 
+utils.event.actions.declare.setMutationDefaults(({ canonicalMutationFn }) => ({
+  // This retry ensures on page reload if event have not yet synced,
+  // the action will be retried once
+  retry: 1,
+  retryDelay: 5000,
+  mutationFn: async (actionInput) => {
+    /*
+     * Handles action being called with an outdated temporary id
+     *
+     * If you have created the event offline and you've also created the action
+     * offline, then once you come back online it might just happen that you have
+     * an action request in the buffer that still uses the temporary id to refer to the event
+     *
+     * This piece of code updates that id before the call is made
+     */
+    const events = await readEventsFromStorage()
+    const eventThatHadATemporaryId = events.find(
+      (e) => e.transactionId === actionInput.eventId
+    )
+
+    if (!eventThatHadATemporaryId) {
+      return canonicalMutationFn(actionInput)
+    }
+
+    return canonicalMutationFn({
+      ...actionInput,
+      eventId: eventThatHadATemporaryId.id
+    })
+  },
+  onMutate: async (actionInput) => {
+    await queryClient.cancelQueries({ queryKey: ['persisted-events'] })
+    const events = await readEventsFromStorage()
+    const eventToUpdate = events.find(
+      (e) =>
+        e.id === actionInput.eventId ||
+        // This hook is executed before mutationFn, so we need to check for both ids
+        e.transactionId === actionInput.eventId
+    )!
+
+    const eventsWithoutUpdated = events.filter(
+      (e) => e.id !== actionInput.eventId
+    )
+    const previousActions = eventToUpdate.actions
+    await writeEventToStorage([...eventsWithoutUpdated, eventToUpdate])
+    queryClient.invalidateQueries({ queryKey: ['persisted-events'] })
+    return { previousActions, events }
+  },
+  onSettled: async (response) => {
+    /*
+     * Updates event in store
+     */
+    if (response) {
+      await queryClient.cancelQueries({ queryKey: ['persisted-events'] })
+      const events = await readEventsFromStorage()
+      const eventsWithoutNew = events.filter((e) => e.id !== response.id)
+
+      await writeEventToStorage([...eventsWithoutNew, response])
+      return queryClient.invalidateQueries({ queryKey: ['persisted-events'] })
+    }
+  }
+}))
+
 utils.event.create.setMutationDefaults(({ canonicalMutationFn }) => ({
-  mutationFn: canonicalMutationFn
+  mutationFn: canonicalMutationFn,
+  retry: 0,
+  onMutate: async (newEvent) => {
+    const optimisticEvent = {
+      id: newEvent.transactionId,
+      type: newEvent.type,
+      transactionId: newEvent.transactionId,
+      createdAt: new Date(),
```

Comment:
Should we opt for toISOString()  with dates? easier to manage types end to end.

Reply:
If I understand correctly, `superjson` transforms the Date to ISO string before sending and then transforms it back when backend receives it